internal/proxy/request_signer.go

```diff
@@ -0,0 +1,176 @@
+package proxy
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"hash"
+	"io/ioutil"
+	"net/http"
+	"strings"
+)
+
+// Only headers enumerated in this list are used to compute the signature of a request.
+var signedHeaders = []string{
+	"Content-Length",
+	"Content-Md5",
+	"Content-Type",
+	"Date",
+	"Authorization",
+	"X-Forwarded-User",
+	"X-Forwarded-Email",
+	"X-Forwarded-Groups",
+	"Cookie",
+}
+
+// Name of the header used to transmit the signature computed for the request.
+var signatureHeader = "Octoboi-Signature"
+
+// RequestSigner exposes an interface for digitally signing requests using an RSA private key.
+// See comments for the Sign() method below, for more on how this signature is constructed.
+type RequestSigner struct {
+	hasher       hash.Hash
+	signingKey   crypto.Signer
+	publicKeyStr string
+}
+
+// NewRequestSigner constructs a RequestSigner object from a PEM+PKCS8 encoded RSA public key.
+func NewRequestSigner(signingKeyPemStr string) (*RequestSigner, error) {
+	var privateKey crypto.Signer
+	var publicKey string
+
+	// Build private key.
+	if block, _ := pem.Decode([]byte(signingKeyPemStr)); block == nil {
+		return nil, fmt.Errorf("could not read PEM block from signing key")
+	} else if key, err := x509.ParsePKCS8PrivateKey(block.Bytes); err != nil {
+		return nil, fmt.Errorf("could not read key from signing key bytes: %s", err)
+	} else {
+		privateKey = key.(crypto.Signer)
+	}
+
+	// Derive public key.
+	rsaPublicKey, ok := privateKey.Public().(*rsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("only RSA public keys are currently supported")
+	}
+	publicKey = string(pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "RSA PUBLIC KEY",
+			Bytes: x509.MarshalPKCS1PublicKey(rsaPublicKey),
+		}))
+
+	return &RequestSigner{
+		hasher:       sha256.New(),
+		signingKey:   privateKey,
+		publicKeyStr: publicKey,
+	}, nil
+}
+
+// mapRequestToHashInput returns a string representation of a Request, formatted as a
+// newline-separated sequence of entries from the request. Any two Requests sharing the same
+// representation are considered "equivalent" for purposes of verifying the integrity of a request.
+//
+// Representations are formatted as follows:
+//   <HEADER.1>
+//   ...
+//   <HEADER.N>
+//   <URL>
+//   <BODY>
+//  where:
+//    <HEADER.k> is the ','-joined concatenation of all header values of `signedHeaders[k]`; all
+//      other headers in the request are ignored,
+//    <URL> is the string "<PATH>(?<QUERY>)(#FRAGMENT)", where "?<QUERY>" and "#<FRAGMENT>" are
+//      ommitted if the associated components are absent from the request URL,
+//    <BODY> is the body of the Request (may be `nil`; e.g. for GET requests).
```

Comment:
I have some concerns about signing the body -- particularly if individuals are trying to upload large content files. As it exists today, I believe sso proxy will gracefully handle these cases by streaming the body and not loading them entirely into memory prior to proxying them. 

I think we have to break that behavior here in to support the signing of the body, which would entail us loading the body into memory before proxying the request. 

How important is it to sign the body? Are there other mechanisms we can use to compensate for these risks?

Reply:
Can we configure signing requests on a per-upstream basis and have it on by default, but for services that may have request bodies that are too large, turn off request signing?

Reply:
@jphines - Currently, the HMAC signing library that we use reads the whole request body into memory, and factors it into the hash. Having said that, I think that the risk you identify is legitimate.

@shrayolacrayon - I think that's a smart idea.